### PR TITLE
report: add label to provide context for subItems

### PR DESCRIPTION
**Pull request by connorjclark**
This adds a field `label` to `TableSubItems`, to help improve context when rendering subitems underneath a table item row.

|      Audit      |  subitem label    |
| --------------- | ----------------- |
| axe-audit (base) | Related Elements |
| color-contrast | Background Element |
| duplicate-id-aria | Duplicate Elements |
| duplicated-javascript | Script |
| legacy-javascript | Location |
| unused-javascript | Module |
| csp-xss | Reasons |
| inspector-issues | URL |
| non-composited-animations | Reasons |
| hreflang | Reasons |
| third-party-facades | Script |

image

image

image

_____

Did not give `third-party-web` a subitems label because it seems good already

image

cnn report here: https://lighthouse-1065v6gec-googlechrome.vercel.app/gh-pages/viewer/?gist=8fdec0793bab3abea4e66bc1129b06d7

### CODE REVIEW

**Comment by connorjclark** [comment]
`hr` too much?


light mode

image

**Comment by adamraine** [comment]
> hr too much?

IMO yes. Would underline work?

**Comment by connorjclark** [comment]
wb some `width: max-content` on the `lh-text` to limit the size to just under the text?

image


underline looks pretty weaksauce imo:

image

**Comment by adamraine** [comment]
Yeah underline is weaksauce, bad idea. Even the shortened line looks a little weird.

I played around a bit, I think the full line would work better if it were gray instead of white/black:

Screen Shot 2022-04-15 at 3 59 11 PM

File: types/lhr/audit-details.d.ts [hunk 1]
```diff
@@ -117,6 +117,7 @@ declare module Details {
   interface TableSubItems {
     type: 'subitems';
     items: TableItem[];
+    label?: string | IcuMessage;
```
Comment (brendankenny) [inline types/lhr/audit-details.d.ts:120]: The duplication of the label for every group of subItems is unfortunate. What about defining this at a higher level (like a property of the table itself) and we can add overriding at the `TableSubItems` level if/when we need it?

File: report/renderer/details-renderer.js [hunk 2]
```diff
@@ -424,7 +424,21 @@ export class DetailsRenderer {
...
+      const rowEl = this._dom.createElement('tr');
+      const tdEl = this._dom.createChildOf(rowEl, 'td');
+      tdEl.append(this._renderText(item.subItems.label));
+      tdEl.append(this._dom.createElement('hr'));
+      rowEl.classList.add('lh-sub-item-row');
+      // Need a `td` per column so that background color of the table will
+      // span the width of the label row.
+      for (let i = 1; i < headings.length; i++) {
+        this._dom.createChildOf(rowEl, 'td');
+      }
```
Comment (brendankenny) [inline report/renderer/details-renderer.js:439]: you're looking for [`colspan`](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/td#attr-colspan) :)

File: lighthouse-core/test/audits/byte-efficiency/legacy-javascript-test.js [hunk 3]
```diff
@@ -98,28 +98,33 @@ describe('LegacyJavaScript audit', () => {
     ]);
     expect(result.items).toHaveLength(1);
     expect(result.items[0]).toMatchInlineSnapshot(`
+Object {
```
Comment (brendankenny) [inline lighthouse-core/test/audits/byte-efficiency/legacy-javascript-test.js:101]: indentation issue with these files?

File: lighthouse-core/audits/accessibility/axe-audit.js [hunk 4]
```diff
@@ -16,11 +16,17 @@ const i18n = require('../../lib/i18n/i18n.js');
 const UIStrings = {
   /** Label of a table column that identifies HTML elements that have failed an audit. */
   failingElementsHeader: 'Failing Elements',
+  /** Label of a table column that identifies HTML elements that are related to a failure in an audit. */
+  relatedElementsHeader: 'Related Elements',
```
Comment (paulirish) [inline lighthouse-core/audits/accessibility/axe-audit.js:20]:
```suggestion
  relatedElementsHeader: 'Related elements',
```

File: lighthouse-core/audits/accessibility/color-contrast.js [hunk 5]
```diff
@@ -23,6 +23,8 @@ const UIStrings = {
   /** Description of a Lighthouse audit that tells the user *why* they should try to pass. This is displayed after a user expands the section to see more. No character length limits. 'Learn More' becomes link text to additional documentation. */
   description: 'Low-contrast text is difficult or impossible for many users to read. ' +
       '[Learn more](https://web.dev/color-contrast/).',
+  /** Label of a table column that identifies a single HTML element that is the background of another HTML element. */
+  backgroundElementHeader: 'Background Element',
```
Comment (paulirish) [inline lighthouse-core/audits/accessibility/color-contrast.js:27]:
```suggestion
  backgroundElementHeader: 'Background element',
```

File: lighthouse-core/audits/accessibility/duplicate-id-aria.js [hunk 6]
```diff
@@ -20,6 +20,8 @@ const UIStrings = {
   failureTitle: 'ARIA IDs are not unique',
   /** Description of a Lighthouse audit that tells the user *why* they should try to pass. This is displayed after a user expands the section to see more. No character length limits. 'Learn More' becomes link text to additional documentation. */
   description: 'The value of an ARIA ID must be unique to prevent other instances from being overlooked by assistive technologies. [Learn more](https://web.dev/duplicate-id-aria/).',
+  /** Label of a table column that identifies HTML elements that are duplicates of another HTML element. */
+  relatedElementsHeader: 'Duplicate Elements',
```
Comment (paulirish) [inline lighthouse-core/audits/accessibility/duplicate-id-aria.js:24]:
```suggestion
  relatedElementsHeader: 'Duplicate elements',
```

**Review by brendankenny — COMMENTED** [review]
overall I like it better, but the audits with screenshots maybe need another iteration. The screenshot makes it harder to differentiate the levels and they also get busier with the extra label and hr

File: lighthouse-core/test/audits/byte-efficiency/legacy-javascript-test.js (hunk 3 above)
Comment (connorjclark) [reply lighthouse-core/test/audits/byte-efficiency/legacy-javascript-test.js:101]: oops. You have to babysit jest when asking it to `-u` inline snapshots. It was fixed for a short minute... but regressed at some point. Maybe fixed in latest, idk.

File: lighthouse-core/audits/accessibility/axe-audit.js [hunk 7]
```diff
@@ -16,11 +16,17 @@ const i18n = require('../../lib/i18n/i18n.js');
...
 };
 
 const str_ = i18n.createMessageInstanceIdFn(__filename, UIStrings);
 
 class AxeAudit extends Audit {
+  static get relatedNodesLabel() {
```
[the remark this answers is not in the document]
Comment (connorjclark) [reply lighthouse-core/audits/accessibility/axe-audit.js:26]: Well, this does match how we handle `meta`.

**Comment by paulirish** [comment]
> but the audits with screenshots maybe need another iteration.

yeah.. i tried out some styles to shrink the screenshot + constrain the text for subrows.  it's.. better.

<details><summary>screenshot and whack-ass styles</summary>

image



```css

.lh-sub-item-row .lh-element-screenshot {
  zoom: 0.4
}

.lh-sub-item-row .lh-node {
  max-height: calc(100px * 0.4);
  overflow: hidden;
  display: block;

}
```

</details>

**Review by paulirish — COMMENTED** [review]
these capitals feel extraneous. (though perhaps we adopted a style for this already?)

i don't have a great idea on the label's typography or graphical delineation...

@jiwoong can you help us out?  

1) we're adding this label to communicate what these sub items are all about. it looks aiight but you could probably help us with it
2) especially when theres more thumbnails in the sub items, the layout gets pretty wonky and hard to follow.. especially tricky cuz we autosize based on aspect ratio and the aspect ratios are totally unpredictable. in an above comment i tried shrinking subitem thumbnails.  any ideas?

File: types/lhr/audit-details.d.ts (hunk 1 above)
Comment (connorjclark) [reply types/lhr/audit-details.d.ts:120]: It seems likely this will come up for inspector-issues one day, but yeah we could move this up to the Table

image


but would you prefer the extra complexity to (probably, one day) support inheritance vs the extra few bytes in the LHR?

File: lighthouse-core/audits/accessibility/axe-audit.js (hunk 4 above)
Comment (connorjclark) [reply lighthouse-core/audits/accessibility/axe-audit.js:20]: yeah but `Failed Elements` :(

File: report/renderer/details-renderer.js (hunk 2 above)
Comment (connorjclark) [reply report/renderer/details-renderer.js:439]: hell yeah, thanks

File: lighthouse-core/audits/accessibility/axe-audit.js (hunk 4 above)
Comment (connorjclark) [reply lighthouse-core/audits/accessibility/axe-audit.js:20]: bump

**Comment by connorjclark** [comment]
made the line thinner (pictures have been updated)

**Comment by connorjclark** [comment]
we don't like how we show the sub-row label multiple times

let's add the label to be just below the main header label, but like this:

```
Failing Element
   | Duplicated Elements
```

and lets also add that vertical line to the sub row items (like already done for insight audit tables)

**Comment by Samiullahstanikzai** [comment]
i always come and fix the issue sometime i forgot to leave to comments and i really apology and I'm thankful to be here :)
